test/compiler/inference.jl

```diff
@@ -1727,6 +1729,90 @@ for expr25261 in opt25261[i:end]
 end
 @test foundslot
 
+# some tests below only work for functions defined in toplevel, but not for closures
+macro evaltoplevel(ex)
+    m = Core.eval(__module__, :(module $(gensym()) end))::Module
+    return QuoteNode(Core.eval(m, ex))
+end
+
+@testset "inter-procedural conditional constraint propagation" begin
+    # simple cases
+    isaint(a) = isa(a, Int)
+    @test Base.return_types((Any,)) do a
+        isaint(a) && return a # a::Int
+        return 0
+    end == Any[Int]
+    eqnothing(a) = a === nothing
+    @test Base.return_types((Union{Nothing,Int},)) do a
+        eqnothing(a) && return 0
+        return a # a::Int
+    end == Any[Int]
+
+    # more complicated cases
+    ispositive(a) = isa(a, Int) && a > 0
+    @test Base.return_types((Any,)) do a
+        ispositive(a) && return a # a::Int
+        return 0
+    end == Any[Int]
+    @test (@evaltoplevel begin
+        isaint2(a::Int)           = true
+        isaint2(@nospecialize(_)) = false
+
+        Base.return_types((Any,)) do a
+            isaint2(a) && return a # a::Int
+            return 0
+        end
+    end) == Any[Int]
+    @test (@evaltoplevel begin
+        ispositive2(a::Int) = a > 0
+        ispositive2(a)      = false
+        Base.return_types((Any,)) do a
+            ispositive2(a) && return a # a::Int
+            return 0
+        end
+    end) == Any[Int]
+
+    # type constraints from multiple constant boolean return types
+    @test (@evaltoplevel begin
+        function f(x)
+            isa(x, Int) && return true
+            isa(x, Symbol) && return true
+            return false
+        end
+        Base.return_types((Any,)) do x
+            f(x) && return x # ::Union{Int,Symbol}
+            return nothing
+        end
+    end) == Any[Union{Int,Symbol,Nothing}]
+
+    # with Base functions
+    @test Base.return_types((Any,)) do a
+        Base.Fix2(isa, Int)(a) && return a # a::Int
+        return 0
+    end == Any[Int]
+    @test Base.return_types((Union{Nothing,Int},)) do a
+        isnothing(a) && return 0
+        return a # a::Int
+    end == Any[Int]
+    @test Base.return_types((Union{Missing,Int},)) do a
+        ismissing(a) && return 0
+        return a # a::Int
+    end == Any[Int]
+    @test Base.return_types((Any,)) do x
+        Meta.isexpr(x, :call) && return x # x::Expr
+        return nothing
+    end == Any[Union{Nothing,Expr}]
+
+    # TODO: back-propagate multiple conditional constraints
+    @test_broken (@evaltoplevel begin
+        is_int_and_int(a, b) = isa(a, Int) && isa(b, Int)
+        Base.return_types((Any,Any)) do a, b
+            is_int_and_int(a, b) && return a, b # (a::Int, b::Int) ideally, but (a::Any, b::Int)
+            0, 0
+        end
+    end) == Any[Tuple{Int,Int}]
```

Comment:
I changed the old `isexpr` test case to this one for the future multiple conditional constraint back-propagation enhancement.